File: ios/ExpoKakaoModule.swift
import ExpoModulesCore
import KakaoSDKCommon
import KakaoSDKAuth
import KakaoSDKUser

struct KakaoAuthError: Error {
    let reason: String
}

public class ExpoKakaoModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoKakao")

    Constants([
      "PI": Double.pi
    ])

    AsyncFunction("initializeKakaoSDK") { (appKey: String, options: [String: Any]?) -> Void in
    }

    Function("isKakaoTalkLoginUrl") { (url: String) -> Bool in
      guard let urlObj = URL(string: url),
            let _ = try? KakaoSDK.shared.appKey() else { return false }
      return AuthApi.isKakaoTalkLoginUrl(urlObj)
    }

    Function("handleOpenUrl") { (url: String) -> Bool in
      guard let urlObj = URL(string: url),
            let _ = try? KakaoSDK.shared.appKey() else { return false }
      return AuthController.handleOpenUrl(url: urlObj)
    }

    AsyncFunction("isKakaoTalkLoginAvailable") { () -> Bool in
      return UserApi.isKakaoTalkLoginAvailable()
    }

    AsyncFunction("login") { (options: [String: Any]) -> [String: Any] in
      // 옵션을 파싱하여 타입에 맞게 변환
      let serviceTerms = options["serviceTerms"] as? [String] ?? []
      let prompts = options["prompts"] as? [String]
      let useKakaoAccountLogin = options["useKakaoAccountLogin"] as? Bool ?? false
      let scopes = options["scopes"] as? [String] ?? []

      return try await withCheckedThrowingContinuation { continuation in
        let callback: (OAuthToken?, Error?) -> Void = { token, error in
          if let error = error {
            continuation.resume(throwing: KakaoAuthError(reason: error.localizedDescription))
          } else if let token = token {
            continuation.resume(returning: self.tokenToDictionary(token))
          } else {
            continuation.resume(throwing: KakaoAuthError(reason: "Unknown error occurred"))
          }
        }

        // scopes가 비어있지 않은 경우
        if !scopes.isEmpty {
          UserApi.shared.loginWithKakaoAccount(scopes: scopes, completion: callback)
        } 
        // 카카오톡 앱을 통한 로그인 가능하고, useKakaoAccountLogin이 false인 경우
        else if UserApi.isKakaoTalkLoginAvailable(), !useKakaoAccountLogin {
          UserApi.shared.loginWithKakaoTalk(
            serviceTerms: self.emptyArrayToNil(serviceTerms),
            completion: callback
          )
        } 
        // 기본 카카오 계정 로그인을 사용하는 경우
        else {
          UserApi.shared.loginWithKakaoAccount(
            prompts: self.convertPrompts(prompts),
            serviceTerms: self.emptyArrayToNil(serviceTerms),
            completion: callback
          )
        }
      }
    }

    AsyncFunction("me") { () -> [String: Any] in
      return try await withCheckedThrowingContinuation { continuation in
        UserApi.shared.me { user, error in
          if let error = error {
            continuation.resume(throwing: KakaoAuthError(reason: error.localizedDescription))
          } else if let user = user {
            continuation.resume(returning: self.userToDictionary(user))
          } else {
            continuation.resume(throwing: KakaoAuthError(reason: "Unknown error occurred"))
          }
        }
      }
    }
  }

  // MARK: - Helper Methods

  private func emptyArrayToNil<T>(_ arr: [T]?) -> [T]? {
    return (arr?.isEmpty == true) ? nil : arr
  }

  private func convertPrompts(_ prompts: [String]?) -> [Prompt] {
    return prompts?.compactMap { prompt in
        switch prompt {
        case "Login": return .Login
        case "Cert": return .Cert
        case "Create": return .Create
        case "UnifyDaum": return .UnifyDaum
        case "SelectAccount": return .SelectAccount
        default: return nil
        }
    } ?? []
}
  private func tokenToDictionary(_ token: OAuthToken) -> [String: Any] {
    return [
      "accessToken": token.accessToken,
      "refreshToken": token.refreshToken,
      "tokenType": token.tokenType,
      "idToken": token.idToken as Any,
      "accessTokenExpiresAt": token.expiredAt.timeIntervalSince1970,
      "refreshTokenExpiresAt": token.refreshTokenExpiredAt.timeIntervalSince1970,
      "accessTokenExpiresIn": token.expiresIn,
      "refreshTokenExpiresIn": token.refreshTokenExpiresIn,
      "scopes": token.scopes ?? []
    ]
  }

  private func userToDictionary(_ user: User) -> [String: Any] {
    return [
      "id": user.id as Any,
      "name": user.kakaoAccount?.name as Any,
      "email": user.kakaoAccount?.email as Any,
      "nickname": user.kakaoAccount?.profile?.nickname as Any,
      "profileImageUrl": user.kakaoAccount?.profile?.profileImageUrl?.absoluteString as Any,
      "thumbnailImageUrl": user.kakaoAccount?.profile?.thumbnailImageUrl?.absoluteString as Any,
      "phoneNumber": user.kakaoAccount?.phoneNumber as Any,
      "ageRange": user.kakaoAccount?.ageRange?.rawValue as Any,
      "birthday": user.kakaoAccount?.birthday as Any,
      "birthdayType": user.kakaoAccount?.birthdayType as Any,
      "birthyear": user.kakaoAccount?.birthyear as Any,
      "gender": user.kakaoAccount?.gender?.rawValue as Any,
      "isEmailValid": user.kakaoAccount?.isEmailValid as Any,
      "isEmailVerified": user.kakaoAccount?.isEmailVerified as Any
    ]
  }
}

File: android/src/main/java/expo/modules/kakao/ExpoKakaoModule.kt
package expo.modules.kakao

import android.content.Context
import expo.modules.core.errors.ModuleNotFoundException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.kakao.sdk.common.KakaoSdk
import com.kakao.sdk.common.util.Utility
import com.kakao.sdk.user.UserApiClient
import com.kakao.sdk.auth.model.Prompt
import com.kakao.sdk.user.model.User
import com.kakao.sdk.auth.model.OAuthToken
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import java.util.Date
import expo.modules.kotlin.Promise


class ExpoKakaoModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoKakao")

        Constants(
            "PI" to Math.PI
        )

        Function("initializeKakaoSDK") { appKey: String, options: Map<String, Any?> ->
            val context = getValidContext()
            KakaoSdk.init(context, appKey)
            println("Kakao SDK initialized with appKey: $appKey")
        }

        AsyncFunction("isKakaoTalkLoginAvailable") {
            val context = getValidContext()
            UserApiClient.instance.isKakaoTalkLoginAvailable(context)
        }

        Function("getKeyHashAndroid") { ->
            val context = getValidContext()
            Utility.getKeyHash(context)
        }

        AsyncFunction("login") { 
            options: Map<String, Any?>,
            promise: Promise ->

            val serviceTerms = options["serviceTerms"] as? List<String>
            val prompts = options["prompts"] as? List<String>
            val useKakaoAccountLogin = options["useKakaoAccountLogin"] as? Boolean ?: false
            val scopes = options["scopes"] as? List<String>

            val context = appContext.currentActivity
                ?: run {
                    promise.reject("LOGIN_ERROR", "Current activity is not available", null)
                    return@AsyncFunction
                }

            val callback: (OAuthToken?, Throwable?) -> Unit = { token, error ->
                if (error != null) {
                    promise.reject("LOGIN_ERROR", error.message, error)
                } else if (token == null) {
                    promise.reject("LOGIN_ERROR", "Login failed: No token received", null)
                } else {
                    promise.resolve(tokenToMap(token))
                }
            }


            // Kakao SDK의 로그인 방식 선택
            when {
                scopes?.isNotEmpty() == true -> {
                    UserApiClient.instance.loginWithNewScopes(context, scopes, callback = callback)
                }

                UserApiClient.instance.isKakaoTalkLoginAvailable(context) && !useKakaoAccountLogin -> {
                    UserApiClient.instance.loginWithKakaoTalk(
                        context,
                        serviceTerms = serviceTerms,
                        callback = callback
                    )
                }

                else -> {
                    UserApiClient.instance.loginWithKakaoAccount(
                        context,
                        convertPrompts(prompts),
                        serviceTerms = serviceTerms,
                        callback = callback
                    )
                }
            }
        }

        AsyncFunction("me") { promise: Promise ->
            val context = getValidContext()

            UserApiClient.instance.me { user, error ->
                if (error != null) {
                    // Ensure all required parameters are passed
                    promise.reject("USER_INFO_ERROR", error.message ?: "An error occurred", error)
                } else if (user == null) {
                    promise.reject("USER_INFO_ERROR", "Failed to fetch user info", null)
                } else {
                    promise.resolve(userToMap(user))
                }
            }
        }
    }

    private fun tokenToMap(token: OAuthToken): Map<String, Any> {
        return mapOf(
            "accessToken" to (token.accessToken ?: ""),
            "refreshToken" to (token.refreshToken ?: ""),
            "tokenType" to "",
            "idToken" to (token.idToken ?: ""),
            "accessTokenExpiresAt" to token.accessTokenExpiresAt.time.toDouble(),
            "refreshTokenExpiresAt" to token.refreshTokenExpiresAt.time.toDouble(),
            "accessTokenExpiresIn" to ((token.accessTokenExpiresAt.time - Date().time) / 1000).toDouble(),
            "refreshTokenExpiresIn" to ((token.refreshTokenExpiresAt.time - Date().time) / 1000).toDouble(),
            "scopes" to (token.scopes ?: emptyList())
        )
    }

    private fun userToMap(user: User): Map<String, Any?> {
        return mapOf(
            "id" to (user.id ?: ""),
            "name" to (user.kakaoAccount?.name ?: ""),
            "email" to (user.kakaoAccount?.email ?: ""),
            "nickname" to (user.kakaoAccount?.profile?.nickname ?: ""),
            "profileImageUrl" to (user.kakaoAccount?.profile?.profileImageUrl ?: ""),
            "thumbnailImageUrl" to (user.kakaoAccount?.profile?.thumbnailImageUrl ?: ""),
            "phoneNumber" to (user.kakaoAccount?.phoneNumber ?: ""),
            "ageRange" to (user.kakaoAccount?.ageRange?.name ?: ""),
            "birthday" to (user.kakaoAccount?.birthday ?: ""),
            "birthdayType" to (user.kakaoAccount?.birthdayType?.name ?: ""),
            "birthyear" to (user.kakaoAccount?.birthyear ?: ""),
            "gender" to (user.kakaoAccount?.gender?.name ?: ""),
            "isEmailValid" to (user.kakaoAccount?.isEmailValid ?: false),
            "isEmailVerified" to (user.kakaoAccount?.isEmailVerified ?: false)
        )
    }

    private fun convertPrompts(prompts: List<String>?): List<Prompt>? {
        return prompts?.mapNotNull {
            when (it.lowercase()) {
                "login" -> Prompt.LOGIN
                "create" -> Prompt.CREATE
                "cert" -> Prompt.CERT
                "unifydaum" -> Prompt.UNIFY_DAUM
                "selectaccount" -> Prompt.SELECT_ACCOUNT
                else -> null
            }
        }
    }

    private fun getValidContext(): Context {
        return appContext.reactContext
            ?: throw IllegalStateException("ReactContext is not available")
    }

}
